# Pandas Challenge Solutions! 

# Part A: Load in the data

Turns out, you actually have the data on your computer! It lives in `jtc_class_code/datasets/billboard_1978.csv`

1. First, make a new folder called `27_pandas` inside your `jtc_class_code/challenges` folder
    * Then make a new jupyter notebook inside this folder called `pandas_challenge.ipynb`
2. Load it in using `pd.read_csv()` and save it into a variable called songs_1978. 
    * **NOTE:** Since the data file is in a different folder, you'll need to use the `../` syntax to specify exactly where your data live

In [4]:
# import pandas
import pandas as pd
df = pd.read_csv('../../datasets/billboard_1978.csv')

# Part B: Inspect the data 

Take a look at the whole DataFrame
### 3. Use `head()` and `tail()` to check out the first and last few rows. Which song ranked #94 on the chart?

In [24]:
df.head()

Year  Hot_Singles_Rank                   Title       Artist    Genre  \
0  1978                 1        "Shadow Dancing"    Andy Gibb      Pop   
1  1978                 2           "Night Fever"     Bee Gees      Pop   
2  1978                 3  "You Light Up My Life"  Debby Boone  Country   
3  1978                 4         "Stayin' Alive"     Bee Gees      Pop   
4  1978                 5     "Kiss You All Over"        Exile  Country   

  Weeks_On_Chart  Peak_Chart_Position Maj_Min  BPM  
0             25                    1     Min  102  
1             20                    1     Min  110  
2             25                    1     Maj   78  
3             27                    1     Min  103  
4             23                    1     Maj  104

In [9]:
df.tail(7)

Year  Hot_Singles_Rank                     Title  \
93  1978                94           "Turn to Stone"   
94  1978                95  "I Can't Stand the Rain"   
95  1978                96              "Ebony Eyes"   
96  1978                97    "The Name of the Game"   
97  1978                98         "We're All Alone"   
98  1978                99        "Hollywood Nights"   
99  1978               100            "Deacon Blues"   

                                Artist Genre Weeks_On_Chart  \
93            Electric Light Orchestra  Rock             15   
94                            Eruption   Pop             22   
95                           Bob Welch  Rock             17   
96                                ABBA   Pop             16   
97                       Rita Coolidge   Pop             20   
98  Bob Seger & The Silver Bullet Band  Rock             13   
99                          Steely Dan  Rock             16   

    Peak_Chart_Position Maj_Min  BPM  
93                   13     Maj  138  
94                   18     Maj  108  
95                   14     Maj  118  
96                   12     Maj   76  
97                    7     Maj   66  
98                   12     Maj  150  
99                   19     Maj  114

'Turn to stone' is ranked 94 on the chart

### 4. How many rows and columns are in this DataFrame?

In [11]:
df.shape

(100, 9)

100 rows, 9 columns

### 5. Which columns have numeric (integer or float) data? 

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Year                 100 non-null    int64 
 1   Hot_Singles_Rank     100 non-null    int64 
 2   Title                100 non-null    object
 3   Artist               100 non-null    object
 4   Genre                100 non-null    object
 5   Weeks_On_Chart       100 non-null    object
 6   Peak_Chart_Position  100 non-null    int64 
 7   Maj_Min              100 non-null    object
 8   BPM                  100 non-null    int64 
dtypes: int64(4), object(5)
memory usage: 7.2+ KB


Numeric data in `Year`, `Hot_Singles_Rank`, `Peak_Chart_Position`, `BPM`

### 6. Which columns have strings (listed as 'object' when you run `info()`)?

`Title`, `Artist`, `Genre`, `Weeks_On_Chart`, `Maj_Min`

### 7. What is slowest tempo (minimum BPM) and fastest tempo (maximum BPM) this year?

In [20]:
df.describe()

Year  Hot_Singles_Rank  Peak_Chart_Position         BPM
count   100.0        100.000000           100.000000  100.000000
mean   1978.0         50.500000             6.940000  107.700000
std       0.0         29.011492             5.112512   20.396821
min    1978.0          1.000000             1.000000   64.000000
25%    1978.0         25.750000             3.000000   96.000000
50%    1978.0         50.500000             6.000000  110.000000
75%    1978.0         75.250000            11.000000  122.000000
max    1978.0        100.000000            21.000000  156.000000

Slowest tempo is 64 and fastest is 156 beats per minute

# Part C: Access specific cells of the dataframe

### 8. The 2nd ranking song in 1978 was called 'Night Fever'. Print out the contents of the cell storing the artist for that song using `.iloc()`
    
    

In [40]:
# the second row (index 1) and the 4th column (index 3)
df.iloc[1,3]

'Bee Gees'

#### Now, print out the contents of the cell storing the genre for that song


In [41]:
# this column is one more to the right (the 5th column), so it is index 4
df.iloc[1,4]

'Pop'

### 9. Print out the name of song ranking 50 on the chart 
    * (hint: which row of the dataframe will this be?)

In [42]:
# tricky because the 50th song is at row index 49 since indexing starts at 0
df.iloc[49, 2]

'"Thunder Island"'